### DISCUSSION TITLE
PWM "Hi-Fi" mode taken from Mozzi

### DISCUSSION THREAD

**Discussion by donluca** [probonopd/MiniDexed#440] [Ideas]
[The Mozzi library](https://github.com/sensorium/Mozzi) offers on some boards the possibility of using a ["Hi-Fi" mode by combining 2 PWM signals](https://sensorium.github.io/Mozzi/doc/html/group__core.html#gae99eb43cb29bb03d862ae829999916c4/).

I'm not sure how this is achieved, but since the original DX7 had a mono output, I think it would be reasonable to offer the possibility to the user of giving up stereo output to have better quality output through PWM.

[The output circuit to sum the two separate PWM channels](https://sensorium.github.io/Mozzi/learn/output/) is really simple and requires just 3 extra components (2 resistors and a cap).

EDIT: [here is a good explanation of how this works](https://www.openmusiclabs.com/learning/digital/pwm-dac/dual-pwm-circuits/)

**Comment by diyelectromusic** [comment]
The problem is that the Raspberry Pi includes circuitry to produce the PWM output on the headphone jacks, and that circuitry is set up for a certain resolution, stereo signal. It will require a hardware change to swap the stereo signal for a higher resolution mono signal using the technique as used in Mozzi.  See: https://hackaday.com/2018/07/13/behind-the-pin-how-the-raspberry-pi-gets-its-audio/ for an explanation, and here for the same circuit on a Pi 4: https://datasheets.raspberrypi.com/rpi4/raspberry-pi-4-reduced-schematics.pdf

It might be possible to produce independent PWM audio using GPIO directly using the type of setup described for Mozzi, but then it won't be using the on-board 3.5mm jack socket...

... and if we start adding extra circuitry to GPIO, then it really is a lot simpler to use a cheap DAC :)

Kevin

**Reply by donluca** [reply]
Fair point, although I have to point out that soldering a couple of resistors and a cap to a 3.5mm socket is not a huge task and definitely cheaper than buying a DAC board.
Not to mention that every hobbyist probably already has those basic components on hand, so no need to even buy those. You could even desolder the built in 3.5mm socket from the board itself and reuse it :P

**Reply by diyelectromusic** [reply]
It's not about adding components to the 3.5mm socket, it's about removing all the filtering that already exists on the Raspberry Pi before the signal gets to the 3.5mm jack - from the circuit diagram, each channel has what appears to be a buffer IC, 3 resistors and two capacitors - at least - that is just the bits I can understand with my limited knowledge :)

I am not an electronics person, so I can't say that a dual PWM signal couldn't be created from the two channels based on the filter circuitry that already exist, but at the very least I seriously doubt the Mozzi method would work "as is" given all the additional circuitry already being applied to the PWM signal off the GPIO...  The Mozzi method assumes no pre-existing filtering on the PWM, and that the PWM output is a straight 5V PWM from an IO pin...

So to gain the benefit of the onboard PWM, I'd say you're basically into cutting tracks and re-patching the Pi at this point.

As I say you could do this with different GPIO pins but then you're changing firmware, configuring pins, adding circuitry using IO that you probably want to use for something else, although again arguably you're saving pins from not using I2S... but those DACs whilst not as cheap as a few resistors are pretty cheap, and the step change in quality between PWM and a DAC is significant and probably a lot more useful than the step between "some PWM" and "maybe better resolution PWM".  And of course they are basically "hassle free"... so I find it hard to think of a plausible reason why it would justify the effort in the MiniDexed firmware for an inferior output?

The benefit of the onboard 3.5 mm signal PWM is "no additional work or components" at the expense of quality - it allows people to get going really quickly.  The benefit of the DAC is quality at, well, some modest expense.  I struggle to see the use-case for loss of simplicity, still low quality (although not quite as a low as standard PWM), for the sake of not shelling out for a DAC...

So I see the following ways forward if you really want to do it:
* Use different GPIO and implement firmware and a new circuit bypassing the onboard PWM completely.
* Design a new circuit to add to the existing filters in the Pi's PWM output circuit to do what you want.
* Hack the Pi to remove the existing filter circuit and replace it with the "Mozzi method" tailored for a 3V3 PWM output of course.

But none of these really seem worth the tradeoff for the marginal increase in quality you might get from a higher resolution PWM signal compared to just getting that DAC! :)

All my own opinion of course, but at the very least, the required circuit calculations will be a bit more complicated than those in that explanatory link :)

Kevin

**Reply by donluca** [reply]
Your analysis is, obviously, 100% on point.

I never intended cutting traces and patching components onto the existing RPi circuit, rather I was thinking more about using two free pins for PWM duty where we can easily solder the 3 components necessary to have the alternative "high quality" PWM output.

I admit haven't looked into the code at all so this was all more brainstorming rather than a real suggestion and, as a consequence, I have no idea of the amount of effort necessary to pull this off.

I gather from your post that it wouldn't be trivial and, obviously, the quality improvement would be marginal. When you put the two side by side it becomes pretty clear that it wouldn't be worth the effort and, at best, it would be more for curiosity or "academical" reasons, if you wish.

Maybe in the future someone will revisit this and decide to give it a spin. Until then, 3€ dedicated i2s DAC it is. :)
